### ISSUE TITLE
Expose internal operator implementation architecture

### ISSUE THREAD

**Issue by glopesdev** [dotnet/reactive#798]
Internally, the Rx library provides an architecture for performant implementation of Reactive operators based on the `Producer` and `Sink` classes.

While early on it would have made sense to keep this implementation internal to reduce the burden on maintenance, this pattern has now become a source of pressure to keep including new operators inside the core library itself. Some operators that would have otherwise remained outside the library are being considered for inclusion with the argument that there is no other way to implement them performantly. I take this to be a symptom that we need an easier way to provide external Rx library developers with the means to properly implement new operators that are on par with Rx's own internal architecture.

Given that the `Producer`/`Sink` pattern has matured and stabilized over time, is there any reason why we should not expose the pattern to consumers of the library, and document its proper use? Rx encourages composition over reimplementation, which is great and should continue, but time has shown that invariably one will need, for particular reasons, to develop its own operators, and it seems awkward that there is no way to do it performantly.

**Comment by akarnokd** [comment]
> there is no other way to implement them performantly

It's not that bad, the current setup incurst one allocation and one level indirection per observer involved with the operator.

> reduce the burden on maintenance

What I could infer in the past three months is that the current setup is more like due to trust "issues" towards external implementations.

**Comment by danielcweber** [comment]
>What I could infer in the past three months is that the current setup is more like due to trust "issues" towards external implementations.

That's right, the good performance of the internal operators mainly stems from the fact that these are considered somewhat 'safe' wrt. notification grammar, exceptions, disposal etc.

Externally implemented operators will probably always have to be considered less safe, and additional safety measures must be taken (such wrapping observers to become self-muting etc.). I am open to suggestions but I don't see `Producer` and `Sink` being opened to the public but maybe a different approach.

**Comment by glopesdev** [comment]
> Externally implemented operators will probably always have to be considered less safe

I don't understand how this has anything to do with safety. In fact, if anything, I believe it's the opposite: There is no way Rx can guarantee "safeness" of any `IObservable` object. The Rx contract is a set of design guidelines, similar to the .NET design guidelines.

Since the `IObservable` interfaces are public, no one stops me from implementing them directly because of some random performance use case. The chance is extremely high I will do it wrongly, and the consumer of my library will have absolutely no idea unless he digs in the code.

By providing an explicit set of design guidelines and a reference implementation to go with it, I believe you are actually reducing the possibility of unsafe libraries, since you are deliberately teaching people about the issues and tradeoffs in designing reactive operators.

I am not knowledgeable enough in the current implementation details to appreciate the difference between `ObservableBase` and the `Producer`/`Sink` architecture, but if Rx internals are not using `ObservableBase` then I don't think it's a valid solution.

Basically, I reject the claim that nobody else but the core implementors are able to *properly* implement an operator. That is exactly the attitude that got us into a maintenance bottleneck with Rx in the first place.

**Comment by glopesdev** [comment]
> There is no way Rx can guarantee "safeness" of any IObservable object.

I guess that by wrapping external observables you can at least guarantee the basic Rx contract, which is desirable by default.

To be clear, I am not saying that the default should be "trust". I am simply saying that there should be a way for a library developer to say "I know what I'm doing and I'm willing to assume responsibility for any errors".

**Comment by glopesdev** [comment]
Anyway, let's leave it at that for now and resume later whenever there is a need to discuss parity of external libraries.

**Comment by danielcweber** [comment]
>Basically, I reject the claim that nobody else but the core implementors are able to properly implement an operator

That's not what was said, and the discussion should not drift towards 'who implemented what and who is the better expert' and stuff.

Here's the thing: All I can say today about `Producers` and `Sinks` is that they provide some guarantees out af the box (to subscribers) while at the same time being able to assume those guarantees (when subscribing itself to `Producers`). Hence, we're able to gain some performance.

So there's two options:

-Give implementors a means to say 'I know what I'm doing, my observable adheres to the guidelines', while at the same time leaving 'no trust' the default.

-Trust no one, not even internal operators.

As for option 1, IMO that would increase complexity. If operators are buggy although declared otherwise, chances are it'll end up as an issue in this repo.

Option 2 should be explored. Would performance massively decrease?

>I am not knowledgeable enough in the current implementation details to appreciate the difference between ObservableBase and the Producer/Sink architecture, but if Rx internals are not using ObservableBase then I don't think it's a valid solution.

I heartly invite you to explore the differences, check out what `Producers` provide today that `ObservableBase` can't. Where's the bottleneck if operators are implemented with `ObservableBase`? Can it be mitigated somehow?

**Comment by glopesdev** [comment]
> That's not what was said, and the discussion should not drift towards 'who implemented what and who is the better expert' and stuff.

Agreed, I was unnecessarily escalating there.

> I heartly invite you to explore the differences, check out what Producers provide today that ObservableBase can't. Where's the bottleneck if operators are implemented with ObservableBase? Can it be mitigated somehow?

From my initial reading of the source, this seems to agree with @akarnokd that the cost is basically one allocation (`AutoDetachObserver`) and one indirection for each of the forwarding calls per observer. However, looking at the source for `ObserverBase` reveals that there is essentially one `Volatile.Read` call per notification to ensure the observer has not been stopped yet.

I don't think you can do much better than that if you want to guarantee the Rx grammar. One possibility is to mark the `_isStopped` as `volatile` instead of using `Volatile.Read`. Some sources seem to suggest that these are not equivalent and that there may be some small performance benefits depending on processor architecture.

I certainly  wouldn't consider that operation to be negligible, especially on general-purpose high-throughput frameworks such as Rx where you cannot predict how fast things will need to be called.

I think it is desirable to keep moving Rx towards allocation-free paths, and I wouldn't want to regress on the current work that has proven to have very nice performance benefits. I will keep an eye out for possibilities to simplify this.

I agree with you the complexity of the `Producer` / `Sink` architecture is daunting. Too many things can go wrong and the internals and signatures have moved around quite a bit, it is actually much less stable than I thought when I opened the issue, so I do agree there is much more pondering to do before opening up the internals in any way.
